go/vt/vtgate/vstream_manager.go

```diff
@@ -708,49 +744,25 @@ func (vs *vstream) streamFromTablet(ctx context.Context, sgtid *binlogdatapb.Sha
 			err = vterrors.Errorf(vtrpcpb.Code_UNKNOWN, "vstream ended unexpectedly")
 		}
 
-		retry, ignoreTablet := vs.shouldRetry(err)
-		if !retry {
-			log.Errorf("vstream for %s/%s error: %v", sgtid.Keyspace, sgtid.Shard, err)
-			return err
-		}
-		if ignoreTablet {
+		vs.lastError.Record(err)
+		prevErr = err
+
+		if vs.lastError.ShouldRetry() {
```

Comment:
This should be done in the `vstream.shouldRetry` function so that the last error consideration is taking into account along _in addition_ to the other considerations. I think this would likely also address the test failures.